signac/core/synced_collections/synced_collection.py

```diff
@@ -26,6 +28,27 @@
 )
 
 
+class _fake_lock:
+    """A nullary context manager to simulate a lock in backends that don't support them."""
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, type, value, traceback):
+        pass
+
+
+@contextmanager
```

Comment:
Based on [this StackOverflow post](https://stackoverflow.com/questions/26152934/why-the-staggering-overhead-50x-of-contextlib-and-the-with-statement-in-python), I think it would be better to have a class implementing `__enter__` and `__exit__` instead of using `contextmanager` here. I was able to reproduce these benchmarks on my own machine, and the fastest way would look like:
```python
class ManualCatchContext(object):
    def __enter__(self):
        # Acquire lock
        pass

    def __exit__(self, exc_type, exc_val, exc_tb):
        # Release lock
        return True

preinstantiated_manual_catch_context = ManualCatchContext() 
def manual_with_catch_cache(fn):
    with preinstantiated_manual_catch_context:
        fn()
```

Reply:
Yes, I've seen these benchmarks. Did you try and see what effect this has on the overall performance of the code? I'd like to address this in a separate PR that handles context manager performance more generally since we use similar context managers for buffering and suspending synchronization. If contextlib is really a performance bottleneck on the rest of the code then I'd like to just define a simpler contextmanager factory of our own and use it consistently.

Reply:
I haven’t tried this explicitly with the new SyncedCollections code. I _have_ seen this contextmanager overhead appear in the benchmarks I’ve been running for signac-flow. That uses the current implementation and not SyncedCollections, but I suspect the optimization will offer a measurable speedup.

Reply:
Got it. I'll make an issue of this then.